### PR TITLE
Add fromEncoded and fromEncodedArray that accept json string

### PR DESCRIPTION
**Pull request by marto83**
This updates the Message class to include overloads that are needed
when we have encoded messages in json format.

It also adds a check on the CypherEncode to return better error message
when the type of Data is not byte[]

### CODE REVIEW

File: src/IO.Ably.Shared/MessageEncoders/CipherEncoder.cs [hunk 1]
```diff
@@ -33,6 +33,9 @@ public override Result Decode(IMessage payload, ChannelOptions options)
             var cipher = Crypto.GetCipher(options.CipherParams);
             try
             {
+                if((payload.Data is byte[]) == false)
+                    return Result.Fail("Expected data to be byte[] but received " + payload.Data.GetType());
+
```
Comment (withakay) [inline src/IO.Ably.Shared/MessageEncoders/CipherEncoder.cs]: stylecop should not be allowing this style of `if`, there is a rule to always require curly braces.

File: src/IO.Ably.Tests.Shared/Rest/TM3Spec.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                fromEncoded[0].Encoding.Should().BeNullOrEmpty();
+
+                fromEncoded[1].Name.Should().Be("name2");
+                fromEncoded[1].Data.Should().Be("some-data2");
+                fromEncoded[1].Encoding.Should().BeNullOrEmpty();
+            }
+        }
+
+    }
+}
```
Comment (withakay) [inline src/IO.Ably.Tests.Shared/Rest/TM3Spec.cs]: There is a rule that there should always be a carriage return at the end of a file (sorry to nit pick..)

**Comment by withakay** [comment]
Looks good to me, aside from the minor formatting issues.

File: src/IO.Ably.Shared/MessageEncoders/CipherEncoder.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/MessageEncoders/CipherEncoder.cs]: I've updated it. For some reason the full build didn't run so Stylecop didn't pick it up.

File: src/IO.Ably.Tests.Shared/Rest/TM3Spec.cs (hunk 2 above)
Comment (marto83) [reply src/IO.Ably.Tests.Shared/Rest/TM3Spec.cs]: Updated 🛩

**Pull request merged**